### PR TITLE
Prevent sending emails and notifications to inactive users

### PR DESCRIPTION
**Pull request by daviian**
Targets https://github.com/go-gitea/gitea/issues/2034

Added checks if user is inactive (`is_active = false` or `prohibit_login = true`) before creating notifications and sending mails.
Furthermore I added `NOT NULL DEFAULT false` to `prohibit_login` field.

Changed unit tests accordingly and added a migration, which I tested for mssql, mysql and postgresql.

### CODE REVIEW

File: models/issue_mail.go
Comment (lafriks) [inline models/issue_mail.go]: Do not check booleans with `==` do just simple `&& poster.IsActive {`. Should there be also check for `&& !poster.ProhibitLogin`?

File: models/migrations/v39.go
Comment (lafriks) [inline models/migrations/v39.go]: Do not load data in go code. Just update with condition where prohibid_login is null

**Review by lafriks — CHANGES_REQUESTED** [review]

File: models/user.go
Comment (lafriks) [inline models/user.go]: Also add `notnull` here

**Review by lafriks — CHANGES_REQUESTED** [review]

File: models/issue_mail.go
Comment (daviian) [reply models/issue_mail.go]: Yes you are right. Will fix that.

File: models/migrations/v39.go
Comment (daviian) [reply models/migrations/v39.go]: I tried but there was an issue with one database.
Was complaining about not finding anything while updating because I had no user with `null`in database.

Comment (daviian) [reply models/migrations/v39.go]: I tested it again with your proposed change.
It is working now, maybe I had a problem somewhere else ;-)

**Review by lafriks — APPROVED** [review]

File: models/repo_watch.go
Comment (lunny) [inline models/repo_watch.go:56]: I don't think user who is `prohibit_login` will be moved from watchers UI.

Comment (daviian) [reply models/repo_watch.go:56]: The method is also used in https://github.com/go-gitea/gitea/blob/f61a1d210c4ed6f616b4bee25e8231b5bc67a640/models/notification.go#L104

Perhaps I should add a second method or a parameter to add restriction conditionally?

Comment (lafriks) [reply models/repo_watch.go:56]: Yes, add additional parameter to function to know in to return all as it was or just active ones

File: models/issue.go
Comment (ethantkoenig) [inline models/issue.go]: I'm not sure why we're doing a left join here. If the goal is to include comments that don't have a "valid" poster (e.g. deleted user), I don't think this will work; for those comments the `user.is_active` column will be NULL, and the `user.is_active = true` condition won't hold. And if we don't want to include such comments, why not just do an inner join?

Ditto elsewhere

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

File: models/issue.go
Comment (lafriks) [reply models/issue.go]: @ethantkoenig good point, I also think it should be changed to INNER join

**Comment by daviian** [comment]
I will implement the requested changes this week, in the best case this evening.

File: models/repo_watch.go
Comment (daviian) [reply models/repo_watch.go:56]: @lunny @lafriks I figured out that the view isn't using this method anyways. The view is using the method from here https://github.com/daviian/gitea/blob/e3f679f39da7985193cec28a7c21fbc16ab2310d/models/repo_watch.go#L67

The method in question is only used by notification and mail

**Comment by lunny** [comment]
@daviian have you fixed all points?

**Comment by daviian** [comment]
@lunny I have changed the joins to `INNER` as requested by @ethantkoenig and fixed mentioned mistakes from @lafriks 
And I still wait for your answer on my comment regarding the impact on UI.

**Review by ethantkoenig — APPROVED** [review]

File: models/repo_watch.go
Comment (daviian) [reply models/repo_watch.go:56]: @lunny Any response?

**Comment by lafriks** [comment]
LGTM

File: models/repo_watch.go
Comment (lunny) [reply models/repo_watch.go:56]: @daviian OK.

File: models/issue.go
Comment (bkcsoft) [inline models/issue.go]: shouldn't this really be?
```sql
Where("`comment`.issue_id = ?", issue).
```

File: models/issue_watch.go
Comment (bkcsoft) [inline models/issue_watch.go]: Same here, should have `watch.issue_id`

File: models/issue.go
Comment (daviian) [reply models/issue.go]: Since `issue_id` is a unique attribute in this join it is valid sql, but I can add it for the sake of clarity.

File: models/issue_watch.go
Comment (daviian) [reply models/issue_watch.go]: As mentioned in the other comment, I can add it for the sake of clarity, but it wouldn't be necessary.

File: models/migrations/v40.go
Comment (bkcsoft) [inline models/migrations/v40.go]:
```go
_, err = x.Exec(`ALTER TABLE [user] ALTER COLUMN "prohibit_login" BIT NOT NULL`)
```

**Review by bkcsoft — CHANGES_REQUESTED** [review]
One minor thing though

File: models/migrations/v40.go
Comment (daviian) [reply models/migrations/v40.go]: Can I ask why?

Comment (lafriks) [reply models/migrations/v40.go]: No you can't :P I think bkcsoft asked that just for readability to not have to use backslashes

Comment (daviian) [reply models/migrations/v40.go]: So it should be changed in the other statements too?

Comment (ethantkoenig) [reply models/migrations/v40.go]: I would say no, because the other statements contain backticks

**Comment by lafriks** [comment]
@bkcsoft need your approval

**Review by bkcsoft — APPROVED** [review]

**Comment by daviian** [comment]
Rebased with current master

File: models/issue_mail.go
Comment (appleboy) [inline models/issue_mail.go:45]: missing testing in this patch?.

Comment (lafriks) [reply models/issue_mail.go:45]: @appleboy what do you mean?

**Comment by bkcsoft** [comment]
LGTM

**Pull request merged**
